src/main/java/slimeknights/tconstruct/library/traits/ITrait.java

```diff
@@ -0,0 +1,194 @@
+   */
+  float damage(ItemStack tool, LivingEntity player, LivingEntity target, float damage, float newDamage, boolean isCritical);
+
+  /**
+   * Called when an entity is hit, just before the damage is dealt. Damage is the final damage dealt, including critical damage.
+   * Damage has been fully calculated. You can deal damage to the entity in this callback.
+   * The hurtResistantTime will be set correctly before the call, and it will be reset after the call for the original damage call.
+   *
+   * @param tool       The tool dealing the damage.
+   * @param player     The player (or entity) that is hitting the target.
+   * @param target     The entity to hit.
+   * @param damage     The original, unmodified damage from the tool. Does not includes critical damage, that will be calculated afterwards.
+   * @param isCritical If the hit will be a critical hit.
+   */
+  void onHit(ItemStack tool, LivingEntity player, LivingEntity target, float damage, boolean isCritical);
+
+  /**
+   * Modify the knockback applied. Called after onHit and with the actual damage value. Damage value INCLUDES crit damage here.
+   *
+   * @param tool         The tool dealing the damage.
+   * @param player       The player (or entity) that is hitting the target.
+   * @param target       The entity to hit.
+   * @param damage       The damage that will be dealt, including critical hit damage.
+   * @param knockback    Unmodified base knockback
+   * @param newKnockback Current knockback, possibly modified by other traits.
+   * @param isCritical   If the hit will be a critical hit.
+   * @return The knockback, Standard return value is newKnockback.
+   */
+  float knockBack(ItemStack tool, LivingEntity player, LivingEntity target, float damage, float knockback, float newKnockback, boolean isCritical);
+
+  /**
+   * Called after an entity has been hit, after the damage is dealt.
+   *
+   * @param tool        The tool that dealt the damage.
+   * @param player      The player (or entity) that hit the target.
+   * @param target      The entity hit.
+   * @param damageDealt How much damage has been dealt to the entity. This is the ACTUAL damage dealt - the difference in Health of the entity.
+   * @param wasCritical If the hit was a critical hit.
+   * @param wasHit      If the target was actually hit. False when the entity was still invulnerable, or prevented the damage because of some other reason.
+   */
+  void afterHit(ItemStack tool, LivingEntity player, LivingEntity target, float damageDealt, boolean wasCritical, boolean wasHit);
+
+  /** Called when the player holding the tool blocks an attack. */
+  void onBlock(ItemStack tool, PlayerEntity player, LivingHurtEvent event);
+
+  /** Called when the player holding the tool is damaged. Only called if the player is NOT blocking! onBlock is called in that case.*/
+  default void onPlayerHurt(ItemStack tool, PlayerEntity player, LivingEntity attacker, LivingHurtEvent event) {}
```

Comment:
Why are some methods default and some not? I'd almost make all methods default as not all traits use them

Reply:
Leftovers, haven't actually touched the class logic yet. Good point though. I need to check how much of this belongs into abstract/defaults. The only point there is a default is because it was a later addition, so it doesn't break compatibility.